src/pkg/caendr/caendr/services/sql/dataset/_templates.py

```diff
@@ -0,0 +1,163 @@
+        Returns:
+          str: [The URL template associated with the given name.]
+    '''
+    return external_db_url_templates[url_template_type(db_url_name)][db_url_name]
+
+
+def get_url(self, db_url_name: str, species_name: str = None):
+    '''
+      Fills in the specified URL template with version & species information.  Some templates do not require a species.
+        Args:
+          db_url_name (str): [The name of the template to retrieve.]
+          species_name (str, optional): [The species to fill in the template with. Defaults to None. Optional, but must be provided for certain templates.]
+        Raises:
+          BadRequestError: [Class-level WormBase version not set; URL template requires species, but none provided; Unknown species name.]
+        Returns:
+          str: [The desired URL filled in with information for the given species.]
+    '''
+    # Make sure a species name was provided if the URL requires one
+    if (url_template_type(db_url_name) == 'specific') and species_name is None:
+      logger.warning(f'URL template "{db_url_name}" requires a species, but none was provided.')
+      raise BadRequestError()
+
+    # Make sure the species name is valid
+    if species_name is not None and species_name not in self.species_list.keys():
+      logger.warning(f'Cannot construct URL for unknown species "{species_name}".')
+      raise BadRequestError()
+
+    # Get the desired template an fill in species information, if applicable
+    t = Template(get_url_template(db_url_name))
+    if species_name is not None:
+      species = self.get_species(species_name)
+      return t.substitute({
+        'SPECIES': species_name,
+        'PRJ':     species.project_num,
+        'WB':      species.wb_ver,
+      })
+    else:
+      return t.template
+
+
+
+## GS Blobs ##
+
+def get_blob_template(db_url_name: str):
+    '''
+      Gets the blob template associated with a given template name.
+        Args:
+          db_url_name (str): [The name of the template to retrieve.]
+        Returns:
+          str: [The GS blob template associated with the given name.]
+    '''
+    blob_template = internal_db_blob_templates.get(db_url_name)
+    if blob_template is None:
+        logger.warning(f'Unrecognized blob template name "{db_url_name}".')
+        raise BadRequestError()
+    else:
+        return Template(blob_template)
+
+
+def get_blob(self, db_url_name: str, species_name: str):
+
+    # Get the species object
+    species = self.get_species(species_name)
+
+    # Get the blob template
+    t = get_blob_template(db_url_name)
+
+    # Substitute in relevant species values
+    return t.substitute({
+        'SPECIES': species_name,
+        'SVA':     species.sva_ver,
+    })
+
+
+
+## Filenames ##
+
+def get_download_path(self, species_name: str = ''):
+  if species_name == '':
+    return self.local_download_path
+  else:
+    return f'{self.local_download_path}/{species_name}'
```

Comment:
careful -- you need to validate and *only* accept known species_name. Malicious input might set species_name= '../../../../' to access unauthorized locations.

Reply:
Fixed in 3473d8e0c0d73e0daf16cfa3723fbf57223bf9a8